obspy/io/nordic/ellipse.py

```diff
@@ -0,0 +1,329 @@
+        pt1 = self.__relative_viewpoint(pt)
+        x1, y1 = pt1
+        value = ((y1**2)/(self.a**2)) + ((x1**2)/(self.b**2))
+        if debug:
+            print(pt1)
+            print(value)
+        if value <= 1:
+            return True
+        return False
+
+    def is_on(self, pt):
+        """ Is the given point on the ellipse?
+
+        pt = (x,y) coordinates of the point
+        """
+        pt1 = self.__relative_viewpoint(pt)
+        x1, y1 = pt1
+        value = ((y1**2)/(self.a**2)) + ((x1**2)/(self.b**2))
+        if abs(value - 1) < 2*eps:
+            return True
+        return False
+
+    def __relative_viewpoint(self, pt):
+        """ Coordinates of the viewpoint relative to a 'centered' ellipse
+
+        A centered ellipse has its center at 0,0 and its semi-major axis
+        along the y-axis
+
+        Inputs:
+            x,y:   coordinates of the viewpoint
+
+        Outputs:
+            x1,y1: new coordinates of the viewpoint
+        """
+        # Translate
+        pt1 = (pt[0]-self.x, pt[1]-self.y)
+
+        # Rotate
+        R_rot = _ellipse.__ROT_CCW(self.theta)
+        rotated = np.dot(R_rot, pt1)
+        return rotated
+
+    def __absolute_viewpoint(self, pt):
+        """ Coordinates of a point after 'uncentering' ellipse
+
+        Assume that the ellipse was "centered" for calculations, now
+        put the point back in its true position
+
+        Inputs:
+            x,y:   coordinates of the viewpoint
+
+        Outputs:
+            x1,y1: new coordinates of the viewpoint
+        """
+        # Unrotate
+        R_rot = _ellipse.__ROT_CW(self.theta)
+        unrot = np.dot(R_rot, pt)
+        # Untranslate
+        pt1 = (unrot[0] + self.x, unrot[1]+self.y)
+        return pt1
+
+    def __get_tangents(self, pt=(0, 0)):
+        """ Return tangent intersections for a point and the ellipse
+
+        Equation is from http://www.nabla.hr/Z_MemoHU-029.htm
+        P = (-a**2 * m/c, b**2 / c)
+        """
+        if self.is_inside(pt):
+            print('No tangents, point is inside ellipse')
+            return [], []
+        elif self.is_on(pt):
+            print('No tangents, point is on ellipse')
+            return [], []
+
+        # for calculations, assume ellipse is centered at zero and pointing S-N
+        (x1, y1) = self.__relative_viewpoint(pt)
+        ms = np.roots([self.a**2 - y1**2, 2*y1*x1, self.b**2 - x1**2])
+        cs = -(y1 * ms) + x1
+        # Determine the tangent intersect with ellipse
+        T0 = (-self.a**2 * ms[0] / cs[0], self.b**2 / cs[0])
+        T1 = (-self.a**2 * ms[1] / cs[1], self.b**2 / cs[1])
+        T0 = (self.b**2 / cs[0],         -self.a**2 * ms[0] / cs[0])
+        T1 = (self.b**2 / cs[1],         -self.a**2 * ms[1] / cs[1])
+
+        # Rotate back to true coords
+        T0 = self.__absolute_viewpoint(T0)
+        T1 = self.__absolute_viewpoint(T1)
+        return T0, T1
+
+    def subtended_angle(self, pt=(0, 0), debug=False):
```

Comment:
Can you add a test for this?

Reply:
``test_ellipse_subtended_angle()`` added